kotest-assertions/kotest-assertions-shared/src/commonMain/kotlin/io/kotest/assertions/eq/IterableEq.kt

```diff
@@ -62,45 +65,86 @@ object IterableEq : Eq<Iterable<*>> {
       }
    }
 
+   private fun errorWithTypeDetails(actual: Iterable<*>, expected: Iterable<*>): Throwable {
+      val tag = "${actual::class.simpleName?.let {it} ?: actual::class} with ${expected::class.simpleName?.let {it} ?: expected::class}} regardless of content\n"
+      val detailErrorMessage = when {
+         actual is Set<*> || expected is Set<*> -> "Set can be compared only to Set\nCannot compare $tag"
+         actual is Collection || expected is Collection -> "Collection can be compared only to Collection\nCannot compare $tag"
+         else -> "Promiscuous iterators\nCannot compare $tag"
+      }
+      return failure(Expected(expected.print()), Actual(actual.print()), detailErrorMessage)
+   }
+
+   private const val unim = "Unsupported nesting iterator"
+
    private fun checkEquality(actual: Iterable<*>, expected: Iterable<*>, strictNumberEq: Boolean): Throwable? {
       var index = 0
       val iter1 = actual.iterator()
       val iter2 = expected.iterator()
       val elementDifferAtIndex = mutableListOf<Int>()
       var unexpectedElementAtIndex: Int? = null
       var missingElementAt: Int? = null
+      var nestedIteratorError: String? = null
+      var accrueDetails = true
       while (iter1.hasNext()) {
          val a = iter1.next()
          if (iter2.hasNext()) {
             val b = iter2.next()
-            val t = eq(a, b, strictNumberEq)
-            if (t != null) {
+            val t: Throwable? = when {
+               a?.equals(b) == true -> null
+               (a is Iterable<*>) && (a !is Collection<*>) && (a::class.isInstance(actual) || actual::class.isInstance(a)) -> {
```

Comment:
Expression seems hard to read may can be converted to function call.

```kotlin
isNestedIterator(b, expected) -> {
      nestedIteratorError = "$unim $b within $iter2 (must use custom code)"
      accrueDetails = false
      failure(nestedIteratorError)
}
```

```kotlin
private fun isNestedIterator(a: Any?, actual: Iterable<*>) =
      (a is Iterable<*>) && (a !is Collection<*>) && (a::class.isInstance(actual) || actual::class.isInstance(a))
```

Reply:
Converted to function call; I avoided the closure on a `var`, but if you think it's clearer that way, will change so.